analyzers/src/SonarAnalyzer.CSharp/Rules/VariableShadowsField.cs

```diff
@@ -28,75 +28,130 @@ public sealed class VariableShadowsField : SonarDiagnosticAnalyzer
-            context.RegisterNodeAction(c => ProcessVariableDesignation(c, ((VarPatternSyntaxWrapper)c.Node).Designation), SyntaxKindEx.VarPattern);
-            context.RegisterNodeAction(c => ProcessVariableDesignation(c, ((DeclarationPatternSyntaxWrapper)c.Node).Designation), SyntaxKindEx.DeclarationPattern);
-            context.RegisterNodeAction(c => ProcessVariableDesignation(c, ((ListPatternSyntaxWrapper)c.Node).Designation), SyntaxKindEx.ListPattern);
+        private static void Process(SonarSyntaxNodeReportingContext context, SyntaxNode node)
+        {
+            if (ExtractIdentifiers(node) is { Count: > 0 } identifiers
+                && GetContextSymbols(context) is var members)
+            {
+                foreach (var identifier in identifiers)
+                {
+                    ReportOnVariableMatchingField(context, members, identifier);
+                }
+            }
         }
 
-        private static void ProcessVariableDesignation(SonarSyntaxNodeReportingContext context, VariableDesignationSyntaxWrapper variableDesignation)
+        private static List<SyntaxToken> ExtractIdentifiers(SyntaxNode node) =>
+            node switch
+            {
+                VariableDeclarationSyntax variableDeclaration => variableDeclaration.Variables.Select(x => x.Identifier).ToList(),
+                ForEachStatementSyntax foreachStatement => new() { foreachStatement.Identifier },
+                _ when VariableDesignationSyntaxWrapper.IsInstance(node) => ((VariableDesignationSyntaxWrapper)node).AllVariables().Select(x => x.Identifier).ToList(),
+                _ => new()
+            };
+
+        private static List<ISymbol> GetContextSymbols(SonarSyntaxNodeReportingContext context)
         {
-            if (variableDesignation.AllVariables() is { Length: > 0 } variables
-                && context.ContainingSymbol.ContainingType is { } containingType
-                && GetMembers(containingType) is var members)
+            var symbols = context.ContainingSymbol.ContainingSymbol.GetSymbolType().GetMembers();
+            var primaryCtor = symbols.FirstOrDefault(x => x is IMethodSymbol && IsPrimaryCtor(x));
+            var primaryCtorParameters = primaryCtor?.GetParameters();
+            var relevantSymbols = new List<ISymbol>();
+            foreach (var relevant in symbols.Where(x => x is IPropertySymbol or IFieldSymbol))
             {
-                foreach (var variable in variables)
+                if (primaryCtorParameters is not null
+                    && primaryCtorParameters.Any()
+                    && TryGetAssignedValue(relevant, out var assignedValue)
+                    && primaryCtorParameters.Any(x => x.Name == assignedValue))
+                {
+                    primaryCtorParameters = primaryCtorParameters.Where(x => x.Name != assignedValue).ToImmutableArray();
+                }
```

Comment:
I don't think we need this. In this case:
```cs
class C(int i)
{
    private int i = i;
    void M()
    {
        var i = 0; // here the local "i" shadows both "this.i" and the primary constructor parameter "i".
    }
}
```
Here the local variable `i` shadows both the field and the primary constructor parameter. Why do you want to exclude the field and the parameter from the reporting?
I would add both locations (of the field symbol and the parameter symbol) as secondary locations or prioritize any field over any parameter of the same name.

Reply:
Yes you are right, at the beginning I thought about registering for fields and properties as well. In that case, we should not raise if they are assigned to the primary ctor parameter. But I guess that's for another time
```cs
class MyClass(int a)
{
    public int a = a; // Compliant
    public int a; // Noncompliant {{Rename 'a' which hides the primary constructor parameter with the same name.}}
}
```